stetho/src/main/java/com/facebook/stetho/inspector/elements/android/ActivityTracker.java

```diff
@@ -0,0 +1,209 @@
+ * all versions of Android but it is possible to manually track activities using the {@link #add}
+ * and {@link #remove} methods exposed below.  Be aware that this is an easy opportunity to
+ * cause serious memory leaks in your application however.  Use with caution.
+ * <p/>
+ * Most callers can and should ignore this class, though it is necessary if you are implementing
+ * Activity tracking pre-ICS.
+ */
+@NotThreadSafe
+public class ActivityTracker {
+  private static final ActivityTracker sInstance = new ActivityTracker();
+
+  @GuardedBy("Looper.getMainLooper()")
+  private final ArrayList<Activity> mActivities = new ArrayList<>();
+  private final List<Activity> mActivitiesUnmodifiable = Collections.unmodifiableList(mActivities);
+
+  @GuardedBy("ActivityTrackerObservable.mObservers")
+  private final ActivityTrackerObservable mObservable = new ActivityTrackerObservable();
+
+  @Nullable
+  private AutomaticTracker mAutomaticTracker;
+
+  public static ActivityTracker get() {
+    return sInstance;
+  }
+
+  public void registerListener(Listener listener) {
+    mObservable.registerObserver(listener);
+  }
+
+  public void unregisterListener(Listener listener) {
+    mObservable.unregisterObserver(listener);
+  }
+
+  /**
+   * Start automatic tracking if we are running on ICS+.
+   *
+   * @return Automatic tracking has been started.  No need to manually invoke {@link #add} or
+   *     {@link #remove} methods.
+   */
+  public boolean beginTrackingIfPossible(Application application) {
+    if (mAutomaticTracker == null) {
+      AutomaticTracker automaticTracker =
+          AutomaticTracker.newInstance(application, this /* tracker */);
+      if (automaticTracker != null) {
+        automaticTracker.register();
+        mAutomaticTracker = automaticTracker;
+        return true;
+      }
+    }
+    return false;
+  }
+
+  public boolean endTracking() {
+    if (mAutomaticTracker != null) {
+      mAutomaticTracker.unregister();
+      return true;
+    }
+    return false;
+  }
+
+  public void add(Activity activity) {
+    Util.throwIfNull(activity);
+    Util.throwIfNot(Looper.myLooper() == Looper.getMainLooper());
+    mActivities.add(activity);
+    mObservable.notifyActivityAdded(activity);
+  }
+
+  public void remove(Activity activity) {
+    Util.throwIfNull(activity);
+    Util.throwIfNot(Looper.myLooper() == Looper.getMainLooper());
+    if (mActivities.remove(activity)) {
+      mObservable.notifyActivityRemoved(activity);
+    }
+  }
+
+  public List<Activity> getActivitiesView() {
+    return mActivitiesUnmodifiable;
+  }
+
+  private static class ActivityTrackerObservable extends Observable<Listener> {
+    public void notifyActivityAdded(Activity activity) {
+      synchronized (mObservers) {
+        int N = mObservers.size();
+        for (int i = 0; i < N; i++) {
+          Listener listener = mObservers.get(i);
+          listener.onActivityAdded(activity);
```

Comment:
calling foreign code while holding a lock ... hiss! I'm also concerned about re-entrancy here. like if we're notifying clients about onActivityAdded(), during which something else happens and a new Activity launches and we end up back inside ourself. From what I understand, Activities are launched fairly infrequently and there aren't very many of them, so it should be reasonable to copy the list in order to dispatch these notifications to listeners.